File: src/test/js/formulas/math/SIGNTest.ts
import { assert } from "chai";
import { describe, it } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { SIGN } from "../../../../main/js/formulas/math/SIGN";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("SIGN", function () {
  it("should do normal operations", function () {
    assert.equal(SIGN.SELF.run(null, 10), 1);
    assert.equal(SIGN.SELF.run(null, -10), -1);
    assert.equal(SIGN.SELF.run(null, 0), 0);
    assert.equal(SIGN.SELF.run(null, -0), 0);
    assert.equal(SIGN.SELF.run(null, -218637221.22), -1);
    assert.equal(SIGN.SELF.run(null, true), 1);
    assert.equal(SIGN.SELF.run(null, false), 0);
  });

  it("should do string conversion", function () {
    assert.equal(SIGN.SELF.run(null, "10"), 1);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(SIGN.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new SIGN(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(6);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 1);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, -111).add(0, 1, "A").build();
    assert.deepEqual(SIGN.SELF.run(null, one), -1);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((SIGN.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((SIGN.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});

File: src/test/java/io/protobase/f7/formulas/math/SIGNTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class SIGNTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(SIGN.SELF.apply(null, 10.0)).isEqualTo(1.0);
    assertThat(SIGN.SELF.apply(null, -10.0)).isEqualTo(-1.0);
    assertThat(SIGN.SELF.apply(null, 0.0)).isEqualTo(0.0);
    assertThat(SIGN.SELF.apply(null, -0.0)).isEqualTo(0.0);
    assertThat(SIGN.SELF.apply(null, -218637221.22)).isEqualTo(-1.0);
    assertThat(SIGN.SELF.apply(null, true)).isEqualTo(1.0);
    assertThat(SIGN.SELF.apply(null, false)).isEqualTo(0.0);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(SIGN.SELF.apply(null, "10.0")).isEqualTo(1.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(SIGN.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(-1.0);
  }

  @Test
  public void test_lookup() {
    SIGN F = new SIGN(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(1.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(SIGN.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(SIGN.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}
